src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -364,34 +373,262 @@ public void WriteValue(sbyte value)
+
+                if (firstIndexToEscape != -1)
+                {
+                    WriteEscapedStringChunk(chunk, firstIndexToEscape, isFinalBlock, out charsNotProcessedFromPreviousChunk);
+                }
+                else
+                {
+                    WriteStringChunk(chunk, isFinalBlock);
+                }
+
+                // Flush the buffer if needed
+                this.FlushIfBufferThresholdReached();
+            }
+
+            this.bufferWriter.Write(doubleQuote.Slice(0, 1).Span);
+
+            // since we bypass the Utf8JsonWriter, we need to signal to other
+            // Write methods that a separator should be written first
+            CheckIfSeparatorNeeded();
+
+            CheckIfAtManualValueArrayStart();
+        }
+
+        /// <summary>
+        /// Determines if any characters in the provided string span require escaping according to the specified encoder.
+        /// </summary>
+        /// <param name="value">The string span to analyze for characters needing escaping.</param>
+        /// <returns>
+        /// The index of the first character requiring escaping, or -1 if no characters need escaping or if the string span is empty.
+        /// </returns>
+        private unsafe int NeedsEscaping(ReadOnlySpan<char> value)
+        {
+            // Some implementations of JavaScriptEncoder.FindFirstCharacterToEncode may not accept
+            // null pointers and guard against that. Hence, check up-front to return -1.
+            if (value.IsEmpty)
+            {
+                return -1;
+            }
+
+            fixed (char* ptr = value)
+            {
+                return this.encoder.FindFirstCharacterToEncode(ptr, value.Length);
+            }
+        }
+
+        /// <summary>
+        /// Writes an escaped string chunk into the buffer.
+        /// </summary>
+        /// <param name="sourceChunk">The source chunk of the string to be escaped.</param>
+        /// <param name="firstIndexToEscape">The index of the first character that requires escaping.</param>
+        /// <param name="isFinalBlock">Indicates whether the chunk is the final block of the string.</param>
+        /// <remarks>
+        /// This method escapes characters within the provided string chunk that require escaping
+        /// according to the specified encoder. It allocates memory for the escaped string and handles
+        /// buffer management using either stack allocation or array pooling. Once the string is escaped,
+        /// it is written into the buffer. If the encoding operation fails, an exception is thrown. 
+        /// </remarks>
+        private void WriteEscapedStringChunk(ReadOnlySpan<char> sourceChunk, int firstIndexToEscape, bool isFinalBlock, out int preChunkNotProcessed)
+        {
+            char[] valueArray = null;
+            const int StackallocCharThreshold = 128;
+
+            int maxSize = firstIndexToEscape + 6 * (sourceChunk.Length - firstIndexToEscape);

```

Comment:
magic number '6' here needs comments. or same as line 483 to define a const int

Reply:
I also suggest creating a well document constant to store that number.

Reply:
done